Regressions caused by changed of bounds on `Pin::set`

Our CI started failing with the nightly compiler, for a day.

### Code

I tried this code:

```rust
struct Foo {}
impl Foo {
    fn set(&self, _: bool) {}
}

fn main() {
    let foo = Foo {};
    let foo : core::pin::Pin<&Foo> = core::pin::Pin::new(&foo);
    foo.set(true);
}
```

It compiles in stable, but not with the nightly compiler:

```
error[E0277]: the trait bound `&Foo: DerefMut` is not satisfied
    --> src/main.rs:27:9
     |
27   |     foo.set(true);
     |         ^^^ the trait `DerefMut` is not implemented for `&Foo`
     |
note: required by a bound in `Pin::<Ptr>::set`
    --> /home/olivier/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/pin.rs:1478:14
     |
1476 |     pub fn set(&mut self, value: Ptr::Target)
     |            --- required by a bound in this associated function
1477 |     where
1478 |         Ptr: DerefMut,
     |              ^^^^^^^^ required by this bound in `Pin::<Ptr>::set`
help: consider mutably borrowing here
     |
27   |     (&mut foo).set(true);
     |     +++++    +

error[E0308]: mismatched types
    --> src/main.rs:27:13
     |
27   |     foo.set(true);
     |         --- ^^^^ expected `Foo`, found `bool`
     |         |
     |         arguments to this method are incorrect
     |
note: method defined here
    --> /home/olivier/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/pin.rs:1476:12
     |
1476 |     pub fn set(&mut self, value: Ptr::Target)
     |            ^^^

```

### Version it worked on

It works in beta (1.81), it only started failing with nightly very recently (one or two day)

### Version with regression


```
rustc 1.82.0-nightly (c6db1ca3c 2024-08-25)
binary: rustc
commit-hash: c6db1ca3c93ad69692a4c4b5542f26fda4bf3aec
commit-date: 2024-08-25
host: x86_64-unknown-linux-gnu
release: 1.82.0-nightly
LLVM version: 19.1.0
```

Looking at the recent history, I believe the regression is caused by https://github.com/rust-lang/rust/commit/b968b26c03510c80782c4524a334f5fa440cb2d8  (https://github.com/rust-lang/rust/pull/129449)

<!--
If you know when this regression occurred, please add a line like below, replacing `{channel}` with one of stable, beta, or nightly.

@rustbot modify labels: +regression-from-stable-to-{nightly} -regression-untriaged
-->

--- a/library/core/src/pin.rs
+++ b/library/core/src/pin.rs
@@ -1370,7 +1370,15 @@ impl<Ptr: Deref> Pin<Ptr> {
         // SAFETY: see documentation on this function
         unsafe { Pin::new_unchecked(&*self.__pointer) }
     }
+}
 
+// These methods being in a `Ptr: DerefMut` impl block concerns semver stability.
+// Currently, calling e.g. `.set()` on a `Pin<&T>` sees that `Ptr: DerefMut`
+// doesn't hold, and goes to check for a `.set()` method on `T`. But, if the
+// `where Ptr: DerefMut` bound is moved to the method, rustc sees the impl block
+// as a valid candidate, and doesn't go on to check other candidates when it
+// sees that the bound on the method.
+impl<Ptr: DerefMut> Pin<Ptr> {
     /// Gets a mutable reference to the pinned value this `Pin<Ptr>` points to.
     ///
     /// This is a generic method to go from `&mut Pin<Pointer<T>>` to `Pin<&mut T>`.
@@ -1402,10 +1410,7 @@ impl<Ptr: Deref> Pin<Ptr> {
     /// ```
     #[stable(feature = "pin", since = "1.33.0")]
     #[inline(always)]
-    pub fn as_mut(&mut self) -> Pin<&mut Ptr::Target>
-    where
-        Ptr: DerefMut,
-    {
+    pub fn as_mut(&mut self) -> Pin<&mut Ptr::Target> {
         // SAFETY: see documentation on this function
         unsafe { Pin::new_unchecked(&mut *self.__pointer) }
     }
@@ -1420,10 +1425,7 @@ impl<Ptr: Deref> Pin<Ptr> {
     #[unstable(feature = "pin_deref_mut", issue = "86918")]
     #[must_use = "`self` will be dropped if the result is not used"]
     #[inline(always)]
-    pub fn as_deref_mut(self: Pin<&mut Pin<Ptr>>) -> Pin<&mut Ptr::Target>
-    where
-        Ptr: DerefMut,
-    {
+    pub fn as_deref_mut(self: Pin<&mut Pin<Ptr>>) -> Pin<&mut Ptr::Target> {
         // SAFETY: What we're asserting here is that going from
         //
         //     Pin<&mut Pin<Ptr>>
@@ -1475,12 +1477,13 @@ impl<Ptr: Deref> Pin<Ptr> {
     #[inline(always)]
     pub fn set(&mut self, value: Ptr::Target)
     where
-        Ptr: DerefMut,
         Ptr::Target: Sized,
     {
         *(self.__pointer) = value;
     }
+}
 
+impl<Ptr: Deref> Pin<Ptr> {
     /// Unwraps this `Pin<Ptr>`, returning the underlying `Ptr`.
     ///
     /// # Safety

